src/executor/evaluate/expr.rs

```diff
@@ -87,3 +89,78 @@ pub fn between<'a>(
 
     Ok(Evaluated::from(Value::Bool(v)))
 }
+
+pub fn case<'a>(
+    operand: Option<Evaluated<'a>>,
+    when_then: Vec<(Evaluated<'a>, Evaluated<'a>)>,
+    else_result: Option<Evaluated<'a>>,
+) -> Result<Evaluated<'a>> {
+    let (_, then_results) = when_then.iter().cloned().unzip::<_, _, Vec<_>, Vec<_>>();
+
+    let results = match &else_result {
+        Some(result) => [&then_results[..], &[result.to_owned()]].concat(),
+        None => then_results,
+    };
+
+    let result_types = results
+        .iter()
+        .map(|rt| -> Result<i32> {
+            match rt.to_owned().try_into()? {
+                Value::Bool(_) => Ok(1),
+                Value::Interval(_) => Ok(2),
+                Value::Date(_) => Ok(3),
+                Value::I64(_) => Ok(4),
+                Value::F64(_) => Ok(5),
+                Value::Str(_) => Ok(6),
+                Value::Time(_) => Ok(7),
+                Value::Timestamp(_) => Ok(8),
+                Value::Null => Ok(9),
+            }
+        })
+        .collect::<Vec<_>>();
+
+    if result_types.iter().any(|t| t != &result_types[0]) {
+        Err(EvaluateError::UnequalResultTypes("CASE".to_owned()).into())
```

Comment:
You can simplify the code by using `Iterator`.

1. `windows` + `all` or `any`
https://doc.rust-lang.org/std/primitive.slice.html#method.windows
2. `try_fold`

Many other approaches may work.
I'd recommend you to try the first option.

Reply:
fyi. `matches!` - https://doc.rust-lang.org/std/macro.matches.html

Reply:
Oh, I never knew about `windows` and `try_fold`. Thanks for help!